signac/core/collection_api.py

```diff
@@ -0,0 +1,183 @@
+# Copyright (c) 2020 The Regents of the University of Michigan
+# All rights reserved.
+# This software is licensed under the BSD 3-Clause License.
+from contextlib import contextmanager
+from collections import defaultdict
+from abc import abstractmethod
+from abc import ABCMeta
+
+try:
+    import numpy
+    NUMPY = True
+except ImportError:
+    NUMPY = False
+
+try:
+    from collections.abc import Collection
+except ImportError:
+    # Collection does not exist in Python 3.5, only Python 3.6 or newer.
+
+    from collections.abc import Sized, Iterable, Container
+
+    def _check_methods(C, *methods):
```

Comment:
> This entire except block will be removed, so no need to address the comments within it. @csadorf just for your information, I pulled this code directly out of the CPython library source code from the implementation of the Collection class in Python 3.7.

Then I assume that requires an attribution??

Reply:
If we were keeping it then yes, but it's being removed...

Reply:
Let's make sure to always add the correct attributions when code is copied from a different code base even when it is planned to be removed. Reasons are simple: 1. It is already part of the code base now, because every commit is part of the code base, not just the latest. 2. There is always a chance that it is not removed after all (intentional or by accident) in which case this is simply plagiarism.

Reply:
I want to echo @csadorf explicitly on this. All committed code should attribute snippets like this, even if it is temporary.

Reply:
To clarify, I'm not asking about general policy, only about this specific instance. I agree that I wasn't explicit enough when originally suggesting to Vishav, I simply sent him a code snippet along with a link and I should have clearly indicated that we should cite. I'm just not sure what you want to do about it now, given that we a) are removing, and b) squash and merge our PRs. The latter means that this commit actually _won't_ be part of the commit history, so now that we know we are removing it I'm not sure what action you want to take. Should we modify an upstream commit to add in the attribution? That commit and all associated parts of this snippet will be removed by the merge.

Reply:
If we actually make sure that it is removed then no further action is needed.